[Synapse] Update role assignment/definition related cmdlets

sunsw1994 opened the pull request:
**Description**<!--Mandatory-->
<!--Why this PR? What is changed? What is the effect? etc. A high-quality description can accelerate the review process.-->
Update 'az synapse role' commands to support new role base access control feature.  

**Testing Guide**
<!--Example commands with explanations.-->
az synapse role scope -h
az synapse role assignment -h
az synapse role definition -h

**History Notes**
<!--If your PR is not customer-facing, use {Component Name} in the PR title. Otherwise, use [Component Name] to allow our pipeline to add the title as a history note. If you need multiple history notes or would like to overwrite the note from the PR title, please fill in the following templates.-->

[Synapse] az synapse role scope list:  List all scopes synapse supports.
[Synapse] az synapse role assignment create/list/delete: Adding --scope/--item-type/--item arguments to support manage role assignments based on scope.
[Synapse] az synapse role assignment create/list/delete: Adding --assignee-object-id argument,  it will bypass Graph API and uniquely determine principal object instead of deducing principal object using --assignee argument.
[Synapse] BREAKING CHANGE: az synapse role assignment create: Role names at old version are not allowed, Sql Admin, Apache Spark Admin, Workspace Admin
[Synapse] BREAKING CHANGE: az synapse role assignment create: When --assignee argument can't  uniquely determine the principal object, the command will raise error instead of adding a role assignment for the uncertain principal object. 

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

sunsw1994 commented:
/azp run

yonzhan commented:
Synapse

zesluo commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
+    if bool(item) != bool(item_type):
+        raise CLIError('usage error: --item-type STRING --item STRING')
+
+    return _list_role_assignment(cmd, workspace_name, role, assignee or assignee_object_id,
```
_list_role_assignments?

sunsw1994 replied:
Will fix it.

zesluo commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
 # Delete Synapse Role Assignment
-def delete_role_assignment(cmd, workspace_name, ids=None, assignee=None, role=None):
-    client = cf_synapse_client_accesscontrol_factory(cmd.cli_ctx, workspace_name)
+def delete_role_assignment(cmd, workspace_name, ids=None, assignee=None, assignee_object_id=None, role=None,
```
delete_role_assignment_by_id?

sunsw1994 replied:
Cx can give a set of ids to delete not only one role assignment. Besides, user can delete role assignments by assignee or by role. 

delete_role_assignment_by_id is not friendly to cx.

zesluo commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -61,32 +73,96 @@ def delete_role_assignment(cmd, workspace_name, ids=None, assignee=None, role=No
+        # workspaces/{workspaceName}/bigDataPools/{bigDataPoolName}
+        scope = "workspaces/" + workspace_name + "/" + item_type + "/" + item
+    else:
+        scope = "workspaces/" + workspace_name
```
I just want to mention that these two may be not equal from my memory:  1.scope = "workspaces/" + workspace_name 2. scope = null

sunsw1994 replied:
Hi @zesluo  As far as my understand, the default scope is workspaces/{workspace_name}, we can discuss and confirm with @idear1203 .

Thanks for throwing it out.

idear1203 commented on src/azure-cli/azure/cli/command_modules/synapse/constant.py:
```diff
@@ -12,6 +12,9 @@
 AdministratorType = "activeDirectory"
+ITEM_NAME_MAPPING = {'bigDataPools': '{bigDataPoolName}', 'integrationRuntimes': '{integrationRuntimeName}',
+                     'linkedServices': '{linkedServiceName}', 'credentials': '{credentialName}',
+                     'scopePools': '{scopePoolName}'}
```
>'scopePools': '{scopePoolName}' [](start = 21, length = 31)

let's remove scopePools for now.

idear1203 commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
-    role_id = _resolve_role_id(cmd, role, workspace_name)
-    # get object_id
-    object_id = _resolve_object_id(cmd, assignee, fallback_to_object_id=True)
+def list_role_assignments(cmd, workspace_name, role=None, assignee=None, assignee_object_id=None,
```
>assignee=None, assignee_object_id=None, [](start = 58, length = 39)

For what reason we have both `--assignee` and `--assignee-object-id`? Can we just combine them as one parameter?

sunsw1994 replied:
keep same step with 'az role assignment', I also think `--assignee-object-id` can bypass `--assignee` potential issue. 
`--assignee`  parameter needs to invoke graph api which may meeting time delay issue.

idear1203 replied:
Make sense to me.

evelyn-ys replied:
I checked that only `az role assignment create` have additional `--assignee-object-id` with `--assignee-principal-type`. `list` and `delete` don't.

May I know why `az synapse role assignment list/delete` need `--assignee-object-id` without `--assignee-principal-type`?

sunsw1994 replied:
1. `--assignee-principal-type` is an optional parameter for  `az role assignment create` , and its code following `az role assignment` 
https://github.com/Azure/azure-cli/blob/50ccb24528eb36529d1432808f3ff385cc4c04aa/src/azure-cli/azure/cli/command_modules/role/custom.py#L146

list and delete api don't support `--assignee-principal-type` parameter

2. `--assignee-object-id` bypass Graph API, it can be sent to server directly without transforming  `assignee` to its object_id. Server only accepted object id in GUID format. `--assignee-principal-type` is not required. 

3. Cx can still using `--assignee 00000000-0000-0000-0000-000000000000`, adding  `--assignee-object-id`  just provide a way for cx to bypass Graph API invocation in case of insufficient privileges and provide assignee more accurately

idear1203 reviewed (approved):
:shipit:

sunsw1994 commented:
Hi @jsntcy  Could you help to review and merge this PR for Synapse RBAC commands. Thanks a lot.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/synapse/_help.py:
```diff
@@ -901,15 +923,24 @@
   - name: Create a role assignment using service principal name.
     text: |-
         az synapse role assignment create --workspace-name testsynapseworkspace \\
-        --role "Sql Admin" --assignee sp_name
```
does `Sql Admin` still work?

sunsw1994 replied:
No, now Synapse is using new role name. 
Synapse SQL Administrator
Synapse Apache Spark Administrator
...

https://docs.microsoft.com/en-us/azure/synapse-analytics/security/synapse-workspace-synapse-rbac-roles

Juliehzl replied:
In this way, it is breaking change to existing customers and we should try to avoid such breaking. 
To not break customers, you should try to make previous definition work and add deprecation info to redirect to new definition.

idear1203 replied:
Can we just allow this breaking change considering that the module is still in preview? Now Synapse moves to new names, we don't want to allow the old names.

sunsw1994 replied:
Adding breaking change description at History Notes

[Synapse] BREAKING CHANGE: az synapse role assignment create: Role names at old version are not allowed, Sql Admin, Apache Spark Admin, Workspace Admin

Juliehzl replied:
I see. According to the telemetry, there is no much usage for this command as it is in preview. So we could accept the breaking change this time. Please note that we need to avoid breaking change as possible as we can in future design.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/synapse/constant.py:
```diff
@@ -62,3 +64,17 @@ class SqlPoolConnectionClientAuthenticationType(str, Enum):
     ActiveDirectoryIntegrated = 'ADIntegrated'
+
+
+class PrincipalType(str, Enum):
+    user = "User"
+    group = "Group"
+    service_principal = "ServicePrincipal"
+
+
+class ItemType(str, Enum):
+    bigDataPools = "bigDataPools"
+    scopePools = "scopePools"
+    integrationRuntimes = "integrationRuntimes"
+    credentials = "credentials"
+    linkedServices = "linkedServices"
```
why not define in SDK?

sunsw1994 replied:
@Juliehzl  I think this not belong to SDK's scope, we define these item type for CLI using friendly. User can pass item type and CLI code will use this parameter to generate `scope` which is defined in SDK and received by service. 

`PrincipalType` and `ItemType` 's purpose is only to limit the user's input

@idear1203  Do we have a plan to move it to SDK?

idear1203 replied:
Currently we don't have such definitions in SDK. This is a good suggestion.

Juliehzl replied:
I see.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
+def list_role_assignments(cmd, workspace_name, role=None, assignee=None, assignee_object_id=None,
+                          scope=None, item=None, item_type=None):
+    if bool(assignee) and bool(assignee_object_id):
+        raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
```
Consider the following error type
https://github.com/Azure/azure-cli/blob/3bff786b9e337aa35bb4cc64a09e5600a35663f8/src/azure-cli-core/azure/cli/core/azclierror.py#L152-L154

sunsw1994 replied:
Fixed

Juliehzl commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
+    if bool(assignee) and bool(assignee_object_id):
+        raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
+
+    if bool(item) != bool(item_type):
```
same as above

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/synapse/_help.py:
```diff
@@ -919,11 +950,11 @@
     text: |-
         az synapse role assignment delete --workspace-name testsynapseworkspace \\
-        --role "Sql Admin"
+        --role "Synapse Administrator"
```
Why no new examples for deleting role assignments by scope?

sunsw1994 replied:
Will add it.

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/synapse/commands.py:
```diff
@@ -139,8 +139,12 @@ def get_custom_sdk(custom_module, client_factory):
+        client_factory=None)
+
+    synapse_role_definitions_sdk = CliCommandType(
+        operations_tmpl='azure.synapse.accesscontrol.operations#RoleAssignmentsOperations.{}',
```
```suggestion
        operations_tmpl='azure.synapse.accesscontrol.operations#RoleDefinitionsOperations.{}',
```

sunsw1994 replied:
Fixed.

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/synapse/constant.py:
```diff
@@ -62,3 +64,17 @@ class SqlPoolConnectionClientAuthenticationType(str, Enum):
+
+class ItemType(str, Enum):
+    bigDataPools = "bigDataPools"
+    scopePools = "scopePools"
```
Why `scopePools` here since it has been removed from `ITEM_NAME_MAPPING`? Should it be removed too?

sunsw1994 replied:
Thanks for catching that, will remove it.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/synapse/commands.py:
```diff
@@ -375,18 +379,22 @@ def get_custom_sdk(custom_module, client_factory):
         g.custom_show_command('show', 'get_role_definition')
 
+    with self.command_group('synapse role scope', synapse_role_definitions_sdk,
+                            custom_command_type=get_custom_sdk('accesscontrol', None)) as g:
```
You could declare your client factory here and don't need to define in every fuction.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/synapse/commands.py:
```diff
@@ -139,8 +139,12 @@ def get_custom_sdk(custom_module, client_factory):
-        operations_tmpl='azure.synapse.accesscontrol.operations#AccessControlClientOperationsMixin.{}',
+    synapse_role_assignment_sdk = CliCommandType(
+        operations_tmpl='azure.synapse.accesscontrol.operations#RoleAssignmentsOperations.{}',
+        client_factory=None)
```
why your client factory is None?

sunsw1994 replied:
That is a dataplane command, so we need to create client using workspace name which is provided by user from parameter. So the client will be created when CLI received parameter instead of initialization at here. 

For the management commands, client factory is not None.

Juliehzl replied:
Actually it doesn't matter because client factory will take effect during command execution where user parameter is already loaded. 
For `-h`, we are using operation_tmpl to get parameter list. 
You could see `storage blob` related command, which is data plane command and need user input to create a client factory. 
It is nice to have if you can refactor your design but as we are close to release date. I could accept current code now.

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
+    role_id = _resolve_role_id(cmd, role, workspace_name)
+    object_id = _resolve_object_id(cmd, assignee, fallback_to_object_id=True) if resolve_assignee else assignee
+    client = cf_synapse_role_assignments(cmd.cli_ctx, workspace_name)
+    role_assignments = client.list_role_assignments(role_id, object_id, scope).value
+    if scope:
+        role_assignments = [x for x in role_assignments if x.scope == scope]
```
Why do we need to filter the results by scope again since `client.list_role_assignments` already take `scope` parameter?

evelyn-ys replied:
Oh I saw the previous comments and got the reason. Better to keep the comments~

idear1203 replied:
Good catch. I think we should remove the logic, just return raw RP response. @sunsw1994

sunsw1994 replied:
The server side has not supported list assignments by scope. `scope` parameter will be ignored by server, so I add filter at CLI side.

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -4,55 +4,67 @@
+    object_id = _resolve_object_id(cmd, assignee, fallback_to_object_id=True) if resolve_assignee else assignee
+    client = cf_synapse_role_assignments(cmd.cli_ctx, workspace_name)
+    role_assignments = client.list_role_assignments(role_id, object_id, scope).value
+    if scope:
```
```suggestion
    if scope and role_assignments:
```

sunsw1994 replied:
Fixed

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -61,32 +73,96 @@ def delete_role_assignment(cmd, workspace_name, ids=None, assignee=None, role=No
+                                       assignee_principal_type=assignee_principal_type, assignment_id=assignment_id)
+    except Exception as ex:  # pylint: disable=broad-except
+        if _error_caused_by_role_assignment_exists(ex):  # for idempotent
+            return list_role_assignments(cmd, assignee, role, scope)
```
```suggestion
            return list_role_assignments(cmd, workspace_name, role=role, assignee=assignee, assignee_object_id=assignee_object_id, scope=scope, item=item, item_type=item_type)
```

sunsw1994 replied:
Fixed

sunsw1994 commented on src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py:
```diff
@@ -119,6 +127,11 @@ def _resolve_object_id(cmd, assignee, fallback_to_object_id=False):
                            "If the assignee is a principal id, make sure the corresponding principal is created "
                            "with 'az ad sp create --id {assignee}'.".format(assignee=assignee))
 
+        if len(result) > 1:
```
Add breaking change description at History Notes:

[Synapse] BREAKING CHANGE: az synapse role assignment create: When --assignee can't uniquely determine the principal object, the command will raise error instead of adding a role assignment for the uncertain principal object.

sunsw1994 commented:
@evelyn-ys @Juliehzl  Thanks for your comments, these are helpful! I have fixed them and updated history notes with known breaking change information, could you help to review it and merge this PR. Hope we can catch up the latest release.  Thanks a lot！
